src/server/clientiface.cpp

```diff
@@ -343,24 +343,21 @@ void RemoteClient::GetNextBlocks (
 			const bool want_emerge = !block || !block->isGenerated();
 
 			// if the block is already in the emerge queue we don't have to check again
-			if (want_emerge && emerge->isBlockInQueue(p)) {
-				nearest_emerged_d = d;
-				continue;
-			}
-
-			/*
-				Check occlusion cache first.
-			 */
-			if (m_blocks_occ.find(p) != m_blocks_occ.end())
-				continue;
+			if (!want_emerge || !emerge->isBlockInQueue(p)) {
+				/*
+					Check occlusion cache first.
+				 */
+				if (m_blocks_occ.find(p) != m_blocks_occ.end())
+					continue;
 
-			/*
-				Note that we do this even before the block is loaded as this does not depend on its contents.
-			 */
-			if (m_occ_cull &&
-					env->getMap().isBlockOccluded(p * MAP_BLOCKSIZE, cam_pos_nodes, d >= d_cull_opt)) {
-				m_blocks_occ.insert(p);
-				continue;
+				/*
+					Note that we do this even before the block is loaded as this does not depend on its contents.
+				 */
+				if (m_occ_cull &&
+						env->getMap().isBlockOccluded(p * MAP_BLOCKSIZE, cam_pos_nodes, d >= d_cull_opt)) {
+					m_blocks_occ.insert(p);
+					continue;
+				}
 			}
```

Comment:
do we know that doing a pointless `enqueueBlockEmerge` call below here is not performance-relevant?

Reply:
Yep. I measured it. It's not measurable. It's only pointless when the queue is full, and in that case it is just a size check against the various limits.

Alternatively I thought I'd check for queue full in the original if-block, but that quickly looked unwieldy.

Reply:
> It's only pointless when the queue is full

I mean in the `emerge->isBlockInQueue(p) == true` case. or did I miss a side effect?

Reply:
Oh, in that case, it will update the block in the queue as before. That is more correct anyway.
I was worried it might impact performance, but could not measure a perf difference (it's doing an O(log(N)) lookup and setting whatever flags were passed on the result).